### PR TITLE
run all tests

### PR DESCRIPTION
**Pull request by gwenzek**
## What does this PR do?

Currently only tests files with a `if __name__ == '__main__'` will be run.
We want to run all tests in tests/

Example: https://github.com/facebookresearch/fairseq/blob/279796224f7c1e89d1c431a8a3d223b471b36bf9/tests/test_binarizer.py#L121-L123

Since Fairseq installation was broken #4726 but the CI was green I also realize the CI was relying too much on `pip install --editable` and was making it hard to reproduce some of the bugs users would see.
I propose to run CI with a true installation.

This also lead me to find other bugs in setup.py
Fairseq needs pytorch to correctly installed it's binaries, so pytorch should  be installed **before** we look at setup.py.
So I promoted pytorch to a build requirements.
I also bumped the required version of numpy and pytorch so that we fetch version compatible with python 3.8.

### CODE REVIEW

**Comment by gwenzek** [comment]
Note that using this PR I managed to reproduce #4429 in CI: https://github.com/facebookresearch/fairseq/actions/runs/3093135329/jobs/5005173667

```
2022-09-20 20:14:41 | INFO | fairseq.tasks.text_to_speech | Please install tensorboardX: pip install tensorboardX
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/opt/hostedtoolcache/Python/3.9.13/x64/lib/python3.9/site-packages/fairseq/__init__.py", line 33, in <module>
    import fairseq.criterions  # noqa
  File "/opt/hostedtoolcache/Python/3.9.13/x64/lib/python3.9/site-packages/fairseq/criterions/__init__.py", line 36, in <module>
    importlib.import_module("fairseq.criterions." + file_name)
  File "/opt/hostedtoolcache/Python/3.9.13/x64/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "/opt/hostedtoolcache/Python/3.9.13/x64/lib/python3.9/site-packages/fairseq/criterions/ctc.py", line 21, in <module>
    from fairseq.tasks import FairseqTask
  File "/opt/hostedtoolcache/Python/3.9.13/x64/lib/python3.9/site-packages/fairseq/tasks/__init__.py", line 136, in <module>
    import_tasks(tasks_dir, "fairseq.tasks")
  File "/opt/hostedtoolcache/Python/3.9.13/x64/lib/python3.9/site-packages/fairseq/tasks/__init__.py", line 117, in import_tasks
    importlib.import_module(namespace + "." + task_name)
  File "/opt/hostedtoolcache/Python/3.9.13/x64/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "/opt/hostedtoolcache/Python/3.9.13/x64/lib/python3.9/site-packages/fairseq/tasks/online_backtranslation.py", line 34, in <module>
    from fairseq.sequence_generator import SequenceGenerator
  File "/opt/hostedtoolcache/Python/3.9.13/x64/lib/python3.9/site-packages/fairseq/sequence_generator.py", line 16, in <module>
    from fairseq.models import FairseqIncrementalDecoder
  File "/opt/hostedtoolcache/Python/3.9.13/x64/lib/python3.9/site-packages/fairseq/models/__init__.py", line 235, in <module>
    import_models(models_dir, "fairseq.models")
  File "/opt/hostedtoolcache/Python/3.9.13/x64/lib/python3.9/site-packages/fairseq/models/__init__.py", line 217, in import_models
    importlib.import_module(namespace + "." + model_name)
  File "/opt/hostedtoolcache/Python/3.9.13/x64/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "/opt/hostedtoolcache/Python/3.9.13/x64/lib/python3.9/site-packages/fairseq/models/speech_to_speech/__init__.py", line 7, in <module>
    from .s2s_conformer import *  # noqa
  File "/opt/hostedtoolcache/Python/3.9.13/x64/lib/python3.9/site-packages/fairseq/models/speech_to_speech/s2s_conformer.py", line 13, in <module>
    from fairseq.models.speech_to_speech.s2s_transformer import S2UTTransformerModel
  File "/opt/hostedtoolcache/Python/3.9.13/x64/lib/python3.9/site-packages/fairseq/models/speech_to_speech/s2s_transformer.py", line 22, in <module>
    from fairseq.models.speech_to_text import S2TTransformerEncoder
  File "/opt/hostedtoolcache/Python/3.9.13/x64/lib/python3.9/site-packages/fairseq/models/speech_to_text/__init__.py", line 7, in <module>
    from .convtransformer import *  # noqa
  File "/opt/hostedtoolcache/Python/3.9.13/x64/lib/python3.9/site-packages/fairseq/models/speech_to_text/convtransformer.py", line 23, in <module>
    from fairseq.models.speech_to_text.modules.convolution import infer_conv_output_dim
ModuleNotFoundError: No module named 'fairseq.models.speech_to_text.modules'
Error: Process completed with exit code 1.
```

It only happens when using `pip install fairseq` and not `pip install -e .` and when using another directory so that fairseq isn't part of the path.
I hope this new CI will catch more installation bugs upstream.

File: pyproject.toml
Comment (cbalioglu) [inline pyproject.toml]: According to our documentation we are supporting all PyTorch versions newer than 1.5. Considering v1.5 is very archaic, we can revise that requirement, but I think it would be safer to specify `>=1.10` or `>=1.11`, which are still used pretty widely.

**Review by cbalioglu — COMMENTED** [review]
Neat! Thanks a lot for all fixes! Just left a minor comment; otherwise looks good to me.

File: pyproject.toml
Comment (gwenzek) [reply pyproject.toml]: good point. I've changed it to 1.10, and updated the readme accordingly. I also changed py3.6 to py3.8 since we aren't anymore running 3.6 in CI it will likely break silently at some point anyway.

**Review by cbalioglu — APPROVED** [review]
LGTM!

**Comment by gwenzek** [comment]
I've modified the CI to run on a more recent pytorch release: 1.10.
I also wanted to make it run on 1.12, but the Apex version hardcoded in the config doesn't work there, I'll need to investigate which recent version of Apex work and whether it's actually tested.

**Comment by freewym** [comment]
@gwenzek After this commit was merged, I encountered  an error when doing `pip install --editable .` as below.
I did check that before this commit  the installation is fine under exactly the same environment, and ninja is already installed. Any ideas of what the cause is?
(update: have the same error when doing `pip install .`)

```
2022-10-02 03:28:06 Building wheels for collected packages: fairseq
2022-10-02 03:28:06   Building editable for fairseq (pyproject.toml): started
2022-10-02 03:28:09   Building editable for fairseq (pyproject.toml): finished with status 'error'
2022-10-02 03:28:09   error: subprocess-exited-with-error
2022-10-02 03:28:09   
2022-10-02 03:28:09   × Building editable for fairseq (pyproject.toml) did not run successfully.
2022-10-02 03:28:09   │ exit code: 1
2022-10-02 03:28:09   ╰─> [63 lines of output]
2022-10-02 03:28:09       running editable_wheel
2022-10-02 03:28:09       creating /tmp/pip-wheel-ew6gt3p0/tmpz_li8n9z/fairseq.egg-info
2022-10-02 03:28:09       writing /tmp/pip-wheel-ew6gt3p0/tmpz_li8n9z/fairseq.egg-info/PKG-INFO
2022-10-02 03:28:09       writing dependency_links to /tmp/pip-wheel-ew6gt3p0/tmpz_li8n9z/fairseq.egg-info/dependency_links.txt
2022-10-02 03:28:09       writing entry points to /tmp/pip-wheel-ew6gt3p0/tmpz_li8n9z/fairseq.egg-info/entry_points.txt
2022-10-02 03:28:09       writing requirements to /tmp/pip-wheel-ew6gt3p0/tmpz_li8n9z/fairseq.egg-info/requires.txt
2022-10-02 03:28:09       writing top-level names to /tmp/pip-wheel-ew6gt3p0/tmpz_li8n9z/fairseq.egg-info/top_level.txt
2022-10-02 03:28:09       writing manifest file '/tmp/pip-wheel-ew6gt3p0/tmpz_li8n9z/fairseq.egg-info/SOURCES.txt'
2022-10-02 03:28:09       Traceback (most recent call last):
2022-10-02 03:28:09         File "/miniconda/envs/fairseq/bin/ninja", line 5, in <module>
2022-10-02 03:28:09           from ninja import ninja
2022-10-02 03:28:09       ModuleNotFoundError: No module named 'ninja'
2022-10-02 03:28:09       /tmp/pip-build-env-5gsmf_xe/overlay/lib/python3.8/site-packages/torch/utils/cpp_extension.py:411: UserWarning: Attempted to use ninja as the BuildExtension backend but we could not find ninja.. Falling back to using the slow distutils backend.
2022-10-02 03:28:09         warnings.warn(msg.format('we could not find ninja.'))
2022-10-02 03:28:09       reading manifest file '/tmp/pip-wheel-ew6gt3p0/tmpz_li8n9z/fairseq.egg-info/SOURCES.txt'
2022-10-02 03:28:09       reading manifest template 'MANIFEST.in'
2022-10-02 03:28:09       adding license file 'LICENSE'
2022-10-02 03:28:09       writing manifest file '/tmp/pip-wheel-ew6gt3p0/tmpz_li8n9z/fairseq.egg-info/SOURCES.txt'
2022-10-02 03:28:09       creating '/tmp/pip-wheel-ew6gt3p0/tmpz_li8n9z/fairseq-0.12.2.dist-info'
2022-10-02 03:28:09       adding license file "LICENSE" (matched pattern "LICEN[CS]E*")
2022-10-02 03:28:09       creating /tmp/pip-wheel-ew6gt3p0/tmpz_li8n9z/fairseq-0.12.2.dist-info/WHEEL
2022-10-02 03:28:09       /tmp/pip-build-env-5gsmf_xe/overlay/lib/python3.8/site-packages/setuptools/command/install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
2022-10-02 03:28:09         warnings.warn(
2022-10-02 03:28:09       running build_py
2022-10-02 03:28:09       running build_ext
2022-10-02 03:28:09       Traceback (most recent call last):
2022-10-02 03:28:09         File "/tmp/pip-build-env-5gsmf_xe/overlay/lib/python3.8/site-packages/setuptools/command/editable_wheel.py", line 140, in run
2022-10-02 03:28:09           self._create_wheel_file(bdist_wheel)
2022-10-02 03:28:09         File "/tmp/pip-build-env-5gsmf_xe/overlay/lib/python3.8/site-packages/setuptools/command/editable_wheel.py", line 330, in _create_wheel_file
2022-10-02 03:28:09           files, mapping = self._run_build_commands(dist_name, unpacked, lib, tmp)
2022-10-02 03:28:09         File "/tmp/pip-build-env-5gsmf_xe/overlay/lib/python3.8/site-packages/setuptools/command/editable_wheel.py", line 261, in _run_build_commands
2022-10-02 03:28:09           self._run_build_subcommands()
2022-10-02 03:28:09         File "/tmp/pip-build-env-5gsmf_xe/overlay/lib/python3.8/site-packages/setuptools/command/editable_wheel.py", line 288, in _run_build_subcommands
2022-10-02 03:28:09           self.run_command(name)
2022-10-02 03:28:09         File "/tmp/pip-build-env-5gsmf_xe/overlay/lib/python3.8/site-packages/setuptools/_distutils/cmd.py", line 319, in run_command
2022-10-02 03:28:09           self.distribution.run_command(command)
2022-10-02 03:28:09         File "/tmp/pip-build-env-5gsmf_xe/overlay/lib/python3.8/site-packages/setuptools/dist.py", line 1217, in run_command
2022-10-02 03:28:09           super().run_command(command)
2022-10-02 03:28:09         File "/tmp/pip-build-env-5gsmf_xe/overlay/lib/python3.8/site-packages/setuptools/_distutils/dist.py", line 987, in run_command
2022-10-02 03:28:09           cmd_obj.run()
2022-10-02 03:28:09         File "/tmp/pip-build-env-5gsmf_xe/overlay/lib/python3.8/site-packages/setuptools/command/build_ext.py", line 84, in run
2022-10-02 03:28:09           _build_ext.run(self)
2022-10-02 03:28:09         File "/tmp/pip-build-env-5gsmf_xe/overlay/lib/python3.8/site-packages/Cython/Distutils/old_build_ext.py", line 186, in run
2022-10-02 03:28:09           _build_ext.build_ext.run(self)
2022-10-02 03:28:09         File "/tmp/pip-build-env-5gsmf_xe/overlay/lib/python3.8/site-packages/setuptools/_distutils/command/build_ext.py", line 346, in run
2022-10-02 03:28:09           self.build_extensions()
2022-10-02 03:28:09         File "/tmp/pip-build-env-5gsmf_xe/overlay/lib/python3.8/site-packages/torch/utils/cpp_extension.py", line 434, in build_extensions
2022-10-02 03:28:09           self._check_cuda_version(compiler_name, compiler_version)
2022-10-02 03:28:09         File "/tmp/pip-build-env-5gsmf_xe/overlay/lib/python3.8/site-packages/torch/utils/cpp_extension.py", line 812, in _check_cuda_version
2022-10-02 03:28:09           raise RuntimeError(CUDA_MISMATCH_MESSAGE.format(cuda_str_version, torch.version.cuda))
2022-10-02 03:28:09       RuntimeError:
2022-10-02 03:28:09       The detected CUDA version (11.6) mismatches the version that was used to compile
2022-10-02 03:28:09       PyTorch (10.2). Please make sure to use the same CUDA versions.
2022-10-02 03:28:09       
2022-10-02 03:28:09       error: Support for editable installs via PEP 660 was recently introduced
2022-10-02 03:28:09       in `setuptools`. If you are seeing this error, please report to:
2022-10-02 03:28:09       
2022-10-02 03:28:09       https://github.com/pypa/setuptools/issues
```

**Comment by gwenzek** [comment]
Hi, could you report the pip version you're using ? and torch ? `pip show torch pip`

Also for me the error message is pretty clear, and not related to my commit.

**Comment by freewym** [comment]
@gwenzek Hi, my torch/pip version is:
```
Name: torch
Version: 1.12.1
Summary: Tensors and Dynamic neural networks in Python with strong GPU acceleration
Home-page: https://pytorch.org/
Author: PyTorch Team
Author-email: packages@pytorch.org
License: BSD-3
Location: /miniconda/envs/fairseq/lib/python3.8/site-packages
Requires: typing_extensions
Required-by: torchaudio, torchvision
---
Name: pip
Version: 22.1.2
Summary: The PyPA recommended tool for installing Python packages.
Home-page: https://pip.pypa.io/
Author: The pip developers
Author-email: distutils-sig@python.org
License: MIT
Location: /miniconda/envs/fairseq/lib/python3.8/site-packages
Requires: 
Required-by: 
```
The message is strange because my torch is indeed compiled with CUDA 11.6 (`import torch; torch.version.cuda` gives `11.6`). I have tried the commit right before this PR was merged and there was no such errors.

**Comment by sevensix617** [comment]
@freewym  hi,I have encountered the same problem. Have you solved it？

**Comment by freewym** [comment]
@sevensix617 I haven't. Are you using the same version of PyTorch?

**Comment by gwenzek** [comment]
Hi, I'm sorry, but I have a hard time reproducing.
Here is how I do it:

```
conda create python==3.8 -n fairseq_torch_1_10_3
conda activate fairseq_torch_1_10_3
pip install torch==1.10.1+cu111 torchaudio==0.10.1+cu111 -f https://download.pytorch.org/whl/torch_stable.html
pip install -e .
```

Could someone share a detailed script from a new conda env to the error ?
My guess is that the issue if with the "isolated build" feature of setuptools.
In isolated build, setuptool will create a new env just for building fairseq, which means redownloading a pytorch version which is not the one you have on your system.
So please also try `pip install --no-build-isolation -e .`

**Comment by sevensix617** [comment]
@freewym Yes, I am the same as your torch and cuda versions, and the errors are the same.

**Comment by freewym** [comment]
> Hi, I'm sorry, but I have a hard time reproducing. Here is how I do it:
> 
> ```
> conda create python==3.8 -n fairseq_torch_1_10_3
> conda activate fairseq_torch_1_10_3
> pip install torch==1.10.1+cu111 torchaudio==0.10.1+cu111 -f https://download.pytorch.org/whl/torch_stable.html
> pip install -e .
> ```
> 
> Could someone share a detailed script from a new conda env to the error ? My guess is that the issue if with the "isolated build" feature of setuptools. In isolated build, setuptool will create a new env just for building fairseq, which means redownloading a pytorch version which is not the one you have on your system. So please also try `pip install --no-build-isolation -e .`

I ran the build when building a docker:
```
RUN git clone https://github.com/pytorch/fairseq.git && cd fairseq && \                                                        
        pip install --editable ./ && TORCH_CUDA_ARCH_LIST="3.7;6.0;7.0;8.0;8.6" python setup.py build_ext --inplace
```
after adding "--no-build-isolation" to pip install, pip install seems to work now. However, when running the following `setup.py` command, it has another error when building a cuda lib extension:
```
building 'fairseq.libnat_cuda' extension
        creating /fairseq/build/temp.linux-x86_64-cpython-38/fairseq/clib/libnat_cuda
        Traceback (most recent call last):
          File "<string>", line 2, in <module>
          File "<pip-setuptools-caller>", line 34, in <module>
          File "/fairseq/setup.py", line 252, in <module>
            do_setup(package_data)
          File "/fairseq/setup.py", line 164, in do_setup
            setup(
          File "/miniconda/envs/fairseq/lib/python3.8/site-packages/setuptools/__init__.py", line 87, in setup
            return distutils.core.setup(**attrs)
          File "/miniconda/envs/fairseq/lib/python3.8/site-packages/setuptools/_distutils/core.py", line 185, in setup
            return run_commands(dist)
          File "/miniconda/envs/fairseq/lib/python3.8/site-packages/setuptools/_distutils/core.py", line 201, in run_commands
            dist.run_commands()
          File "/miniconda/envs/fairseq/lib/python3.8/site-packages/setuptools/_distutils/dist.py", line 973, in run_commands
            self.run_command(cmd)
          File "/miniconda/envs/fairseq/lib/python3.8/site-packages/setuptools/dist.py", line 1217, in run_command
            super().run_command(command)
          File "/miniconda/envs/fairseq/lib/python3.8/site-packages/setuptools/_distutils/dist.py", line 992, in run_command
            cmd_obj.run()
          File "/miniconda/envs/fairseq/lib/python3.8/site-packages/setuptools/command/develop.py", line 34, in run
            self.install_for_development()
          File "/miniconda/envs/fairseq/lib/python3.8/site-packages/setuptools/command/develop.py", line 114, in install_for_development
            self.run_command('build_ext')
          File "/miniconda/envs/fairseq/lib/python3.8/site-packages/setuptools/_distutils/cmd.py", line 319, in run_command
            self.distribution.run_command(command)
          File "/miniconda/envs/fairseq/lib/python3.8/site-packages/setuptools/dist.py", line 1217, in run_command
            super().run_command(command)
          File "/miniconda/envs/fairseq/lib/python3.8/site-packages/setuptools/_distutils/dist.py", line 992, in run_command
            cmd_obj.run()
          File "/miniconda/envs/fairseq/lib/python3.8/site-packages/setuptools/command/build_ext.py", line 79, in run
            _build_ext.run(self)
          File "/miniconda/envs/fairseq/lib/python3.8/site-packages/Cython/Distutils/old_build_ext.py", line 186, in run
            _build_ext.build_ext.run(self)
          File "/miniconda/envs/fairseq/lib/python3.8/site-packages/setuptools/_distutils/command/build_ext.py", line 346, in run
            self.build_extensions()
          File "/miniconda/envs/fairseq/lib/python3.8/site-packages/torch/utils/cpp_extension.py", line 765, in build_extensions
            build_ext.build_extensions(self)
          File "/miniconda/envs/fairseq/lib/python3.8/site-packages/Cython/Distutils/old_build_ext.py", line 195, in build_extensions
            _build_ext.build_ext.build_extensions(self)
          File "/miniconda/envs/fairseq/lib/python3.8/site-packages/setuptools/_distutils/command/build_ext.py", line 466, in build_extensions
            self._build_extensions_serial()
          File "/miniconda/envs/fairseq/lib/python3.8/site-packages/setuptools/_distutils/command/build_ext.py", line 492, in _build_extensions_serial
            self.build_extension(ext)
          File "/miniconda/envs/fairseq/lib/python3.8/site-packages/setuptools/command/build_ext.py", line 202, in build_extension
            _build_ext.build_extension(self, ext)
          File "/miniconda/envs/fairseq/lib/python3.8/site-packages/setuptools/_distutils/command/build_ext.py", line 547, in build_extension
            objects = self.compiler.compile(
          File "/miniconda/envs/fairseq/lib/python3.8/site-packages/torch/utils/cpp_extension.py", line 581, in unix_wrap_ninja_compile
            cuda_post_cflags = unix_cuda_flags(cuda_post_cflags)
          File "/miniconda/envs/fairseq/lib/python3.8/site-packages/torch/utils/cpp_extension.py", line 480, in unix_cuda_flags
            cflags + _get_cuda_arch_flags(cflags))
          File "/miniconda/envs/fairseq/lib/python3.8/site-packages/torch/utils/cpp_extension.py", line 1694, in _get_cuda_arch_flags
            arch_list[-1] += '+PTX'
        IndexError: list index out of range
        [end of output]
```
I resolved it by adding `TORCH_CUDA_ARCH_LIST=xxx` before `pip install`.

Do you have any idea of why `--no-build-isolation` is not needed (i.e., has the matched PyTorch version) before? Maybe there is a fix if we know the reason.

**Comment by gwenzek** [comment]
Basically the change I made says that we need torch installed to install fairseq, which was already the case.
Since we can now assume that torch is installed we can directly build the C++ extensions while installing fairseq,
this removes the need to run `python setup.py build_ext --inplace`. 

This allows to use fairseq has a dependency in another project. This is important to avoid needing everyone to commit their research experiment inside fairseq.

The problem with my approach is that when setuptools sees that "torch" is required at build time, it sometime decides to download a new version of pytorch instead of using the one on the current conda env. They call that "build isolation" and they think it's a feature: https://pip.pypa.io/en/stable/reference/build-system/pyproject-toml/#build-isolation
But then if you match code compile for torch 1.12 with a torch 1.10 at runtime you get errors.

But currently there is no way of saying "I want the same version of pytorch during build and install". See https://discuss.python.org/t/support-for-build-and-run-time-dependencies/1513/80 for lengthy discussion about this.
AFAIU The conclusion of the thread is that pip can't fix this, but setuptools or alternative build tools should support that.

So the best workaround I found so far is to disable build isolation so that you're sure to reuse the same pytorch.
The thing I don't get is why I can't reproduce the issue in CI or my machine.

In anycase before my PR:

```
conda install torch torchaudio
git clone https://github.com/facebookresearch/fairseq
cd fairseq
pip install -e .
python setup.py build_ext --inplace
```

Now:

```
conda install torch torchaudio
pip install --no-build-isolation fairseq
```

And the `--no-build-isolation` doesn't seems to be always required, I haven't figured out when setuptools feel like it need to upgrade pytorch.
So I think it's a net win for me using fairseq as a library. But let me know if you feel strongly against that.

**Comment by freewym** [comment]
@gwenzek I see. Thanks for your detailed explanations!

**Pull request merged**
